{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1c634ce6-ed7b-4b69-a001-e113aea97bc9",
   "metadata": {},
   "source": [
    "# Data base creation"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "039eea40-db8a-417a-8208-54c63dd827b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "\n",
    "from ete3 import NCBITaxa"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "458baee5-94b4-4bf1-b57c-b2afacca7bd2",
   "metadata": {},
   "source": [
    "## Helpers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "ebe29000-9c14-40e1-b1e6-a8d41a30767c",
   "metadata": {},
   "outputs": [],
   "source": [
    "def tax_id_translator(tax_id: str):\n",
    "    ncbi = NCBITaxa()\n",
    "    name = ncbi.get_taxid_translator([tax_id])\n",
    "    return name[tax_id]\n",
    "\n",
    "\n",
    "def get_taxonomy(tax_id: str, rank: str):\n",
    "    ncbi = NCBITaxa()\n",
    "    try:\n",
    "        lineage = ncbi.get_lineage(tax_id)\n",
    "        rank_id = [kegit@github.com:AndySECP/MicroBioLLM.gity for key, value in ncbi.get_rank(lineage).items() if value == rank]\n",
    "        return tax_id_translator(rank_id[0])\n",
    "    except:\n",
    "        return \"NA\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddba2f0c-2c4c-4f9a-a197-9c4eb399d7ab",
   "metadata": {},
   "source": [
    "## Assembly summary"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "65f0fcb0-500e-4fe3-8123-7a2ce3da1385",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_49959/1106158931.py:1: DtypeWarning: Columns (34,35,36) have mixed types. Specify dtype option on import or set low_memory=False.\n",
      "  assembly_summary = pd.read_csv(\"assembly_summary_refseq.txt\", skiprows=1, sep= \"\\t\")\n"
     ]
    }
   ],
   "source": [
    "assembly_summary = pd.read_csv(\"assembly_summary_refseq.txt\", skiprows=1, sep= \"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5008b6e-c1a3-47f7-b980-37d7d3a3afdc",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "f837d196-1b2c-45bb-94d6-7a1308aef22c",
   "metadata": {},
   "outputs": [],
   "source": [
    "complete_genome = assembly_summary[assembly_summary[\"assembly_level\"] == \"Complete Genome\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "77bc5962-e8a7-4039-b73e-60fbce895e6b",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_49959/1899025977.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  complete_genome['superkingdom'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"superkingdom\"))\n"
     ]
    }
   ],
   "source": [
    "complete_genome['superkingdom'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"superkingdom\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "5a20c0b1-f92d-40f1-bb84-b4f09458f320",
   "metadata": {},
   "outputs": [],
   "source": [
    "bacteria = complete_genome[complete_genome[\"superkingdom\"] == \"Bacteria\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "4a75a183-01ec-4b1a-9c81-4bfcc57e6a7e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_49959/2337496385.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bacteria['genus'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"genus\"))\n"
     ]
    }
   ],
   "source": [
    "bacteria['genus'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"genus\"))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "9813c5ba-810e-41a6-bc4c-b45e6b053748",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_49959/1269789511.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  bacteria['family'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"family\"))\n"
     ]
    }
   ],
   "source": [
    "bacteria['family'] = complete_genome['taxid'].apply(lambda x: get_taxonomy(str(x), \"family\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "b71c69dc-444f-4390-9e83-bfcc271a7fd8",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Enterobacteriaceae               8318\n",
       "Bacillaceae                      1818\n",
       "Pseudomonadaceae                 1783\n",
       "Staphylococcaceae                1747\n",
       "Streptococcaceae                 1348\n",
       "                                 ... \n",
       "Candidatus Chromulinivoraceae       1\n",
       "Paraconexibacteraceae               1\n",
       "Acetomicrobiaceae                   1\n",
       "Thalassobaculaceae                  1\n",
       "Pleomorphomonadaceae                1\n",
       "Name: family, Length: 516, dtype: int64"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteria['family'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "b0f15a6c-9e58-4695-8b73-bbcf95d37828",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#assembly_accession</th>\n",
       "      <th>bioproject</th>\n",
       "      <th>biosample</th>\n",
       "      <th>wgs_master</th>\n",
       "      <th>refseq_category</th>\n",
       "      <th>taxid</th>\n",
       "      <th>species_taxid</th>\n",
       "      <th>organism_name</th>\n",
       "      <th>infraspecific_name</th>\n",
       "      <th>isolate</th>\n",
       "      <th>...</th>\n",
       "      <th>annotation_provider</th>\n",
       "      <th>annotation_name</th>\n",
       "      <th>annotation_date</th>\n",
       "      <th>total_gene_count</th>\n",
       "      <th>protein_coding_gene_count</th>\n",
       "      <th>non_coding_gene_count</th>\n",
       "      <th>pubmed_id</th>\n",
       "      <th>superkingdom</th>\n",
       "      <th>genus</th>\n",
       "      <th>family</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>56</th>\n",
       "      <td>GCF_000006625.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02604098</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>273119</td>\n",
       "      <td>134821</td>\n",
       "      <td>Ureaplasma parvum serovar 3 str. ATCC 700970</td>\n",
       "      <td>strain=ATCC 700970</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/02/26</td>\n",
       "      <td>655</td>\n",
       "      <td>607</td>\n",
       "      <td>39</td>\n",
       "      <td>11048724</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Ureaplasma</td>\n",
       "      <td>Mycoplasmataceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>57</th>\n",
       "      <td>GCF_000006645.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02604093</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>187410</td>\n",
       "      <td>632</td>\n",
       "      <td>Yersinia pestis KIM10+</td>\n",
       "      <td>strain=KIM10+</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/06/04</td>\n",
       "      <td>4242</td>\n",
       "      <td>3947</td>\n",
       "      <td>105</td>\n",
       "      <td>9826348;12142430</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Yersinia</td>\n",
       "      <td>Yersiniaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59</th>\n",
       "      <td>GCF_000006745.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02603969</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>243277</td>\n",
       "      <td>666</td>\n",
       "      <td>Vibrio cholerae O1 biovar El Tor str. N16961</td>\n",
       "      <td>strain=N16961</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/08/24</td>\n",
       "      <td>3715</td>\n",
       "      <td>3487</td>\n",
       "      <td>127</td>\n",
       "      <td>10952301</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Vibrio</td>\n",
       "      <td>Vibrionaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>60</th>\n",
       "      <td>GCF_000006765.1</td>\n",
       "      <td>PRJNA57945</td>\n",
       "      <td>SAMN02603714</td>\n",
       "      <td>na</td>\n",
       "      <td>reference genome</td>\n",
       "      <td>208964</td>\n",
       "      <td>287</td>\n",
       "      <td>Pseudomonas aeruginosa PAO1</td>\n",
       "      <td>strain=PAO1</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>Annotation submitted by NCBI RefSeq</td>\n",
       "      <td>2019/01/24</td>\n",
       "      <td>5697</td>\n",
       "      <td>5572</td>\n",
       "      <td>106</td>\n",
       "      <td>12644476;12644477;10984043;15716452;15608211;1...</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Pseudomonas</td>\n",
       "      <td>Pseudomonadaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>61</th>\n",
       "      <td>GCF_000006845.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN02604088</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>242231</td>\n",
       "      <td>485</td>\n",
       "      <td>Neisseria gonorrhoeae FA 1090</td>\n",
       "      <td>strain=FA 1090</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/04/10</td>\n",
       "      <td>2267</td>\n",
       "      <td>1967</td>\n",
       "      <td>71</td>\n",
       "      <td>na</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Neisseria</td>\n",
       "      <td>Neisseriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362078</th>\n",
       "      <td>GCF_003006175.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN05908511</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>573</td>\n",
       "      <td>573</td>\n",
       "      <td>Klebsiella pneumoniae</td>\n",
       "      <td>strain=DA48896</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/03/13</td>\n",
       "      <td>5694</td>\n",
       "      <td>5437</td>\n",
       "      <td>123</td>\n",
       "      <td>29220374</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Klebsiella</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362081</th>\n",
       "      <td>GCF_003006055.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN05818722</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1286</td>\n",
       "      <td>1286</td>\n",
       "      <td>Staphylococcus simulans</td>\n",
       "      <td>strain=MR3</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/03/26</td>\n",
       "      <td>2776</td>\n",
       "      <td>2633</td>\n",
       "      <td>84</td>\n",
       "      <td>28743956</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Staphylococcus</td>\n",
       "      <td>Staphylococcaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362120</th>\n",
       "      <td>GCF_002813935.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN06161238</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>90371</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Ty...</td>\n",
       "      <td>strain=FORC50</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2023/12/13</td>\n",
       "      <td>4548</td>\n",
       "      <td>4256</td>\n",
       "      <td>120</td>\n",
       "      <td>na</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362152</th>\n",
       "      <td>GCF_002863765.1</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN03069713</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>46626</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Give</td>\n",
       "      <td>strain=CFSAN024229</td>\n",
       "      <td>na</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/03/05</td>\n",
       "      <td>4865</td>\n",
       "      <td>4604</td>\n",
       "      <td>124</td>\n",
       "      <td>na</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>362154</th>\n",
       "      <td>GCF_000487015.2</td>\n",
       "      <td>PRJNA224116</td>\n",
       "      <td>SAMN01813420</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>1124959</td>\n",
       "      <td>28901</td>\n",
       "      <td>Salmonella enterica subsp. enterica serovar Mu...</td>\n",
       "      <td>strain=420</td>\n",
       "      <td>CFSAN001301</td>\n",
       "      <td>...</td>\n",
       "      <td>NCBI RefSeq</td>\n",
       "      <td>NCBI Prokaryotic Genome Annotation Pipeline (P...</td>\n",
       "      <td>2024/03/05</td>\n",
       "      <td>4647</td>\n",
       "      <td>4401</td>\n",
       "      <td>123</td>\n",
       "      <td>24158624</td>\n",
       "      <td>Bacteria</td>\n",
       "      <td>Salmonella</td>\n",
       "      <td>Enterobacteriaceae</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>37970 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       #assembly_accession   bioproject     biosample wgs_master  \\\n",
       "56         GCF_000006625.1  PRJNA224116  SAMN02604098         na   \n",
       "57         GCF_000006645.1  PRJNA224116  SAMN02604093         na   \n",
       "59         GCF_000006745.1  PRJNA224116  SAMN02603969         na   \n",
       "60         GCF_000006765.1   PRJNA57945  SAMN02603714         na   \n",
       "61         GCF_000006845.1  PRJNA224116  SAMN02604088         na   \n",
       "...                    ...          ...           ...        ...   \n",
       "362078     GCF_003006175.1  PRJNA224116  SAMN05908511         na   \n",
       "362081     GCF_003006055.1  PRJNA224116  SAMN05818722         na   \n",
       "362120     GCF_002813935.1  PRJNA224116  SAMN06161238         na   \n",
       "362152     GCF_002863765.1  PRJNA224116  SAMN03069713         na   \n",
       "362154     GCF_000487015.2  PRJNA224116  SAMN01813420         na   \n",
       "\n",
       "         refseq_category    taxid  species_taxid  \\\n",
       "56                    na   273119         134821   \n",
       "57                    na   187410            632   \n",
       "59                    na   243277            666   \n",
       "60      reference genome   208964            287   \n",
       "61                    na   242231            485   \n",
       "...                  ...      ...            ...   \n",
       "362078                na      573            573   \n",
       "362081                na     1286           1286   \n",
       "362120                na    90371          28901   \n",
       "362152                na    46626          28901   \n",
       "362154                na  1124959          28901   \n",
       "\n",
       "                                            organism_name  infraspecific_name  \\\n",
       "56           Ureaplasma parvum serovar 3 str. ATCC 700970  strain=ATCC 700970   \n",
       "57                                 Yersinia pestis KIM10+       strain=KIM10+   \n",
       "59           Vibrio cholerae O1 biovar El Tor str. N16961       strain=N16961   \n",
       "60                            Pseudomonas aeruginosa PAO1         strain=PAO1   \n",
       "61                          Neisseria gonorrhoeae FA 1090      strain=FA 1090   \n",
       "...                                                   ...                 ...   \n",
       "362078                              Klebsiella pneumoniae      strain=DA48896   \n",
       "362081                            Staphylococcus simulans          strain=MR3   \n",
       "362120  Salmonella enterica subsp. enterica serovar Ty...       strain=FORC50   \n",
       "362152   Salmonella enterica subsp. enterica serovar Give  strain=CFSAN024229   \n",
       "362154  Salmonella enterica subsp. enterica serovar Mu...          strain=420   \n",
       "\n",
       "            isolate  ... annotation_provider  \\\n",
       "56               na  ...         NCBI RefSeq   \n",
       "57               na  ...         NCBI RefSeq   \n",
       "59               na  ...         NCBI RefSeq   \n",
       "60               na  ...         NCBI RefSeq   \n",
       "61               na  ...         NCBI RefSeq   \n",
       "...             ...  ...                 ...   \n",
       "362078           na  ...         NCBI RefSeq   \n",
       "362081           na  ...         NCBI RefSeq   \n",
       "362120           na  ...         NCBI RefSeq   \n",
       "362152           na  ...         NCBI RefSeq   \n",
       "362154  CFSAN001301  ...         NCBI RefSeq   \n",
       "\n",
       "                                          annotation_name annotation_date  \\\n",
       "56      NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/02/26   \n",
       "57      NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/06/04   \n",
       "59      NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/08/24   \n",
       "60                    Annotation submitted by NCBI RefSeq      2019/01/24   \n",
       "61      NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/04/10   \n",
       "...                                                   ...             ...   \n",
       "362078  NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/03/13   \n",
       "362081  NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/03/26   \n",
       "362120  NCBI Prokaryotic Genome Annotation Pipeline (P...      2023/12/13   \n",
       "362152  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/03/05   \n",
       "362154  NCBI Prokaryotic Genome Annotation Pipeline (P...      2024/03/05   \n",
       "\n",
       "       total_gene_count protein_coding_gene_count non_coding_gene_count  \\\n",
       "56                  655                       607                    39   \n",
       "57                 4242                      3947                   105   \n",
       "59                 3715                      3487                   127   \n",
       "60                 5697                      5572                   106   \n",
       "61                 2267                      1967                    71   \n",
       "...                 ...                       ...                   ...   \n",
       "362078             5694                      5437                   123   \n",
       "362081             2776                      2633                    84   \n",
       "362120             4548                      4256                   120   \n",
       "362152             4865                      4604                   124   \n",
       "362154             4647                      4401                   123   \n",
       "\n",
       "                                                pubmed_id superkingdom  \\\n",
       "56                                               11048724     Bacteria   \n",
       "57                                       9826348;12142430     Bacteria   \n",
       "59                                               10952301     Bacteria   \n",
       "60      12644476;12644477;10984043;15716452;15608211;1...     Bacteria   \n",
       "61                                                     na     Bacteria   \n",
       "...                                                   ...          ...   \n",
       "362078                                           29220374     Bacteria   \n",
       "362081                                           28743956     Bacteria   \n",
       "362120                                                 na     Bacteria   \n",
       "362152                                                 na     Bacteria   \n",
       "362154                                           24158624     Bacteria   \n",
       "\n",
       "                 genus              family  \n",
       "56          Ureaplasma    Mycoplasmataceae  \n",
       "57            Yersinia        Yersiniaceae  \n",
       "59              Vibrio        Vibrionaceae  \n",
       "60         Pseudomonas    Pseudomonadaceae  \n",
       "61           Neisseria       Neisseriaceae  \n",
       "...                ...                 ...  \n",
       "362078      Klebsiella  Enterobacteriaceae  \n",
       "362081  Staphylococcus   Staphylococcaceae  \n",
       "362120      Salmonella  Enterobacteriaceae  \n",
       "362152      Salmonella  Enterobacteriaceae  \n",
       "362154      Salmonella  Enterobacteriaceae  \n",
       "\n",
       "[37970 rows x 41 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bacteria"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "63bed0da-cb75-495f-848b-09a91460f3c1",
   "metadata": {},
   "outputs": [],
   "source": [
    "families = ['Enterobacteriaceae', 'Bacillaceae', 'Pseudomonadaceae', 'Staphylococcaceae', 'Streptococcaceae']\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "id": "f384bb3e-a170-4f05-baad-451d09077f19",
   "metadata": {},
   "outputs": [],
   "source": [
    "def select_bacterial_set(families: list, bacteria_df): \n",
    "    selected_genomes = pd.DataFrame()\n",
    "\n",
    "    for family in families:\n",
    "        genomes_family = bacteria_df[bacteria_df['family'] == family].sample(n=100, random_state=42)  # Adjust random_state as needed for reproducibility\n",
    "        selected_genomes = pd.concat([selected_genomes, genomes_family])\n",
    "    return selected_genomes\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "a777bef4-75c3-40d8-a0f9-501cdcb4f418",
   "metadata": {},
   "outputs": [],
   "source": [
    "selected_genomes = select_bacterial_set(families, bacteria)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "eb963d4e-8914-40c4-a03f-4c98472d5c48",
   "metadata": {
    "collapsed": true,
    "jupyter": {
     "outputs_hidden": true
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['GCF_022493475.1',\n",
       " 'GCF_900096795.1',\n",
       " 'GCF_032940055.1',\n",
       " 'GCF_013741155.1',\n",
       " 'GCF_014524605.1',\n",
       " 'GCF_012934455.1',\n",
       " 'GCF_023746725.1',\n",
       " 'GCF_015571875.1',\n",
       " 'GCF_022493615.1',\n",
       " 'GCF_027942215.1',\n",
       " 'GCF_905330215.2',\n",
       " 'GCF_009867035.2',\n",
       " 'GCF_015219805.1',\n",
       " 'GCF_013801705.1',\n",
       " 'GCF_030582455.1',\n",
       " 'GCF_034641715.1',\n",
       " 'GCF_021554655.1',\n",
       " 'GCF_002953095.1',\n",
       " 'GCF_020695605.1',\n",
       " 'GCF_013206965.1',\n",
       " 'GCF_003991155.1',\n",
       " 'GCF_900205295.1',\n",
       " 'GCF_001043215.1',\n",
       " 'GCF_001901405.1',\n",
       " 'GCF_001442495.1',\n",
       " 'GCF_024226085.1',\n",
       " 'GCF_016776125.1',\n",
       " 'GCF_003660125.1',\n",
       " 'GCF_024224355.1',\n",
       " 'GCF_013161625.1',\n",
       " 'GCF_000006925.2',\n",
       " 'GCF_002128365.1',\n",
       " 'GCF_019969235.1',\n",
       " 'GCF_029674705.1',\n",
       " 'GCF_013735895.1',\n",
       " 'GCF_030866865.1',\n",
       " 'GCF_024266785.1',\n",
       " 'GCF_002949515.1',\n",
       " 'GCF_021474325.1',\n",
       " 'GCF_017161345.1',\n",
       " 'GCF_016803875.1',\n",
       " 'GCF_014904185.2',\n",
       " 'GCF_023066645.1',\n",
       " 'GCF_000486405.2',\n",
       " 'GCF_002935105.1',\n",
       " 'GCF_000240905.2',\n",
       " 'GCF_000770155.1',\n",
       " 'GCF_007421865.2',\n",
       " 'GCF_009497055.1',\n",
       " 'GCF_019046905.1',\n",
       " 'GCF_026240875.1',\n",
       " 'GCF_001902195.1',\n",
       " 'GCF_030490045.1',\n",
       " 'GCF_030506115.1',\n",
       " 'GCF_003288375.1',\n",
       " 'GCF_013085025.1',\n",
       " 'GCF_035658275.1',\n",
       " 'GCF_900618295.1',\n",
       " 'GCF_030166415.1',\n",
       " 'GCF_002393365.1',\n",
       " 'GCF_000258365.1',\n",
       " 'GCF_019047765.1',\n",
       " 'GCF_009728895.2',\n",
       " 'GCF_000623655.2',\n",
       " 'GCF_016454485.1',\n",
       " 'GCF_001908875.1',\n",
       " 'GCF_003966795.1',\n",
       " 'GCF_904866475.1',\n",
       " 'GCF_013694425.1',\n",
       " 'GCF_013168095.1',\n",
       " 'GCF_014338425.1',\n",
       " 'GCF_000272775.4',\n",
       " 'GCF_000487255.2',\n",
       " 'GCF_000819645.1',\n",
       " 'GCF_904129595.1',\n",
       " 'GCF_015052765.1',\n",
       " 'GCF_014170015.1',\n",
       " 'GCF_009931235.1',\n",
       " 'GCF_017088425.1',\n",
       " 'GCF_028747905.1',\n",
       " 'GCF_013146845.1',\n",
       " 'GCF_017570625.1',\n",
       " 'GCF_027271235.1',\n",
       " 'GCF_002803805.1',\n",
       " 'GCF_015992265.1',\n",
       " 'GCF_003717795.1',\n",
       " 'GCF_002752975.1',\n",
       " 'GCF_013624225.1',\n",
       " 'GCF_011065365.1',\n",
       " 'GCF_019703045.1',\n",
       " 'GCF_019428605.1',\n",
       " 'GCF_001515725.1',\n",
       " 'GCF_024666505.1',\n",
       " 'GCF_002165115.2',\n",
       " 'GCF_028488865.1',\n",
       " 'GCF_002240335.1',\n",
       " 'GCF_013167715.1',\n",
       " 'GCF_003031775.1',\n",
       " 'GCF_024364905.1',\n",
       " 'GCF_014075595.1',\n",
       " 'GCF_001307805.1',\n",
       " 'GCF_024917135.1',\n",
       " 'GCF_000737305.2',\n",
       " 'GCF_004799565.1',\n",
       " 'GCF_012225885.1',\n",
       " 'GCF_001895885.1',\n",
       " 'GCF_006494795.1',\n",
       " 'GCF_022014755.1',\n",
       " 'GCF_030908605.1',\n",
       " 'GCF_024398395.1',\n",
       " 'GCF_014298115.1',\n",
       " 'GCF_003595605.1',\n",
       " 'GCF_010669045.1',\n",
       " 'GCF_024453785.1',\n",
       " 'GCF_015326725.1',\n",
       " 'GCF_027857035.1',\n",
       " 'GCF_013402775.1',\n",
       " 'GCF_030518635.1',\n",
       " 'GCF_014679335.1',\n",
       " 'GCF_023521615.1',\n",
       " 'GCF_019880325.1',\n",
       " 'GCF_002243325.2',\n",
       " 'GCF_000724485.1',\n",
       " 'GCF_002117165.1',\n",
       " 'GCF_002077215.1',\n",
       " 'GCF_003667885.1',\n",
       " 'GCF_003073375.1',\n",
       " 'GCF_012030675.1',\n",
       " 'GCF_027942435.1',\n",
       " 'GCF_021654915.1',\n",
       " 'GCF_000832385.1',\n",
       " 'GCF_030440255.1',\n",
       " 'GCF_009834085.1',\n",
       " 'GCF_022760025.1',\n",
       " 'GCF_030053815.1',\n",
       " 'GCF_009497015.2',\n",
       " 'GCF_013122275.1',\n",
       " 'GCF_034440835.1',\n",
       " 'GCF_030123085.1',\n",
       " 'GCF_009858215.1',\n",
       " 'GCF_022220505.1',\n",
       " 'GCF_030369915.1',\n",
       " 'GCF_014495825.1',\n",
       " 'GCF_001726125.1',\n",
       " 'GCF_022023815.1',\n",
       " 'GCF_000827045.1',\n",
       " 'GCF_000225265.1',\n",
       " 'GCF_030915485.1',\n",
       " 'GCF_017086565.1',\n",
       " 'GCF_001723585.1',\n",
       " 'GCF_000832785.1',\n",
       " 'GCF_016406185.1',\n",
       " 'GCF_029714185.1',\n",
       " 'GCF_000832505.1',\n",
       " 'GCF_026547205.1',\n",
       " 'GCF_000236605.1',\n",
       " 'GCF_000832465.1',\n",
       " 'GCF_025914055.1',\n",
       " 'GCF_000497485.1',\n",
       " 'GCF_014678925.1',\n",
       " 'GCF_026970855.1',\n",
       " 'GCF_027626015.1',\n",
       " 'GCF_001634305.1',\n",
       " 'GCF_002946455.1',\n",
       " 'GCF_001683235.1',\n",
       " 'GCF_009739945.1',\n",
       " 'GCF_900093775.1',\n",
       " 'GCF_003073255.1',\n",
       " 'GCF_004119575.1',\n",
       " 'GCF_024706255.1',\n",
       " 'GCF_905315685.2',\n",
       " 'GCF_020773035.1',\n",
       " 'GCF_022811845.1',\n",
       " 'GCF_024362715.1',\n",
       " 'GCF_023369715.1',\n",
       " 'GCF_001719225.1',\n",
       " 'GCF_036362945.1',\n",
       " 'GCF_002224345.1',\n",
       " 'GCF_001635915.1',\n",
       " 'GCF_018742245.1',\n",
       " 'GCF_020809125.1',\n",
       " 'GCF_004006475.1',\n",
       " 'GCF_028607045.1',\n",
       " 'GCF_000209795.2',\n",
       " 'GCF_013285325.1',\n",
       " 'GCF_022488425.1',\n",
       " 'GCF_905315045.2',\n",
       " 'GCF_018741165.1',\n",
       " 'GCF_016065635.1',\n",
       " 'GCF_009913555.1',\n",
       " 'GCF_032463625.1',\n",
       " 'GCF_001685435.2',\n",
       " 'GCF_029636965.1',\n",
       " 'GCF_905311405.2',\n",
       " 'GCF_030122905.1',\n",
       " 'GCF_009913475.1',\n",
       " 'GCF_001187595.1',\n",
       " 'GCF_014961825.1',\n",
       " 'GCF_025792455.1',\n",
       " 'GCF_000973605.1',\n",
       " 'GCF_019466095.1',\n",
       " 'GCF_036232285.1',\n",
       " 'GCF_900289125.1',\n",
       " 'GCF_022811805.1',\n",
       " 'GCF_000145825.2',\n",
       " 'GCF_003850585.1',\n",
       " 'GCF_013376775.1',\n",
       " 'GCF_030563585.1',\n",
       " 'GCF_030436015.1',\n",
       " 'GCF_003047145.2',\n",
       " 'GCF_022637915.1',\n",
       " 'GCF_024749085.1',\n",
       " 'GCF_003411785.2',\n",
       " 'GCF_019738995.2',\n",
       " 'GCF_026261625.1',\n",
       " 'GCF_032681185.1',\n",
       " 'GCF_002237425.1',\n",
       " 'GCF_030444675.1',\n",
       " 'GCF_036233655.1',\n",
       " 'GCF_029962205.1',\n",
       " 'GCF_014522265.1',\n",
       " 'GCF_022870785.1',\n",
       " 'GCF_036418275.1',\n",
       " 'GCF_009755645.1',\n",
       " 'GCF_014268495.2',\n",
       " 'GCF_021391435.1',\n",
       " 'GCF_026016325.1',\n",
       " 'GCF_030556675.1',\n",
       " 'GCF_019139675.1',\n",
       " 'GCF_902703215.1',\n",
       " 'GCF_012935715.1',\n",
       " 'GCF_021283465.2',\n",
       " 'GCF_030758435.1',\n",
       " 'GCF_902703195.1',\n",
       " 'GCF_025398015.1',\n",
       " 'GCF_034424585.1',\n",
       " 'GCF_028747325.1',\n",
       " 'GCF_000257545.3',\n",
       " 'GCF_029814935.1',\n",
       " 'GCF_024584495.1',\n",
       " 'GCF_022559545.1',\n",
       " 'GCF_019464415.1',\n",
       " 'GCF_003288335.1',\n",
       " 'GCF_014217705.1',\n",
       " 'GCF_025985525.1',\n",
       " 'GCF_034190815.1',\n",
       " 'GCF_007833895.1',\n",
       " 'GCF_001721745.1',\n",
       " 'GCF_001516265.1',\n",
       " 'GCF_003008615.1',\n",
       " 'GCF_007559125.1',\n",
       " 'GCF_001941865.1',\n",
       " 'GCF_001874645.1',\n",
       " 'GCF_022811045.1',\n",
       " 'GCF_030687475.1',\n",
       " 'GCF_027886605.1',\n",
       " 'GCF_001900265.1',\n",
       " 'GCF_021283365.2',\n",
       " 'GCF_029961665.1',\n",
       " 'GCF_029961625.1',\n",
       " 'GCF_024072255.1',\n",
       " 'GCF_019056475.1',\n",
       " 'GCF_001547955.1',\n",
       " 'GCF_025398835.1',\n",
       " 'GCF_003429185.1',\n",
       " 'GCF_032392015.1',\n",
       " 'GCF_013305685.1',\n",
       " 'GCF_016134935.2',\n",
       " 'GCF_009363155.1',\n",
       " 'GCF_008330565.1',\n",
       " 'GCF_016126955.1',\n",
       " 'GCF_013260315.1',\n",
       " 'GCF_020422905.1',\n",
       " 'GCF_900618265.1',\n",
       " 'GCF_900235835.1',\n",
       " 'GCF_016745135.1',\n",
       " 'GCF_022811025.1',\n",
       " 'GCF_000818015.1',\n",
       " 'GCF_004124255.1',\n",
       " 'GCF_029543085.1',\n",
       " 'GCF_013343555.1',\n",
       " 'GCF_024749345.1',\n",
       " 'GCF_033971305.1',\n",
       " 'GCF_018279285.1',\n",
       " 'GCF_013378035.1',\n",
       " 'GCF_015476275.1',\n",
       " 'GCF_032292255.1',\n",
       " 'GCF_028747405.1',\n",
       " 'GCF_002905815.2',\n",
       " 'GCF_009720405.1',\n",
       " 'GCF_033232085.1',\n",
       " 'GCF_021513395.1',\n",
       " 'GCF_021184005.1',\n",
       " 'GCF_002812825.1',\n",
       " 'GCF_036237015.1',\n",
       " 'GCF_019139875.1',\n",
       " 'GCF_900235825.1',\n",
       " 'GCF_003940825.1',\n",
       " 'GCF_025908315.1',\n",
       " 'GCF_014524625.1',\n",
       " 'GCF_001019485.2',\n",
       " 'GCF_028748305.1',\n",
       " 'GCF_016455085.1',\n",
       " 'GCF_003347055.1',\n",
       " 'GCF_900324065.1',\n",
       " 'GCF_020388765.1',\n",
       " 'GCF_022405375.1',\n",
       " 'GCF_000756715.1',\n",
       " 'GCF_003627735.1',\n",
       " 'GCF_015219905.1',\n",
       " 'GCF_001956755.1',\n",
       " 'GCF_030409015.1',\n",
       " 'GCF_019330185.1',\n",
       " 'GCF_018093205.1',\n",
       " 'GCF_034043415.1',\n",
       " 'GCF_009912935.1',\n",
       " 'GCF_017792285.1',\n",
       " 'GCF_900324315.1',\n",
       " 'GCF_003031485.1',\n",
       " 'GCF_001683315.1',\n",
       " 'GCF_002356175.1',\n",
       " 'GCF_019329545.1',\n",
       " 'GCF_029024305.1',\n",
       " 'GCF_022405295.1',\n",
       " 'GCF_006158985.1',\n",
       " 'GCF_026314335.1',\n",
       " 'GCF_029024945.1',\n",
       " 'GCF_020229655.1',\n",
       " 'GCF_003354805.1',\n",
       " 'GCF_019329505.1',\n",
       " 'GCF_013267575.1',\n",
       " 'GCF_013004085.1',\n",
       " 'GCF_011065325.1',\n",
       " 'GCF_000025085.1',\n",
       " 'GCF_019669385.1',\n",
       " 'GCF_030845335.1',\n",
       " 'GCF_001018655.2',\n",
       " 'GCF_024205825.1',\n",
       " 'GCF_900017775.1',\n",
       " 'GCF_016598895.1',\n",
       " 'GCF_034016775.1',\n",
       " 'GCF_034042975.1',\n",
       " 'GCF_034031975.1',\n",
       " 'GCF_003184985.1',\n",
       " 'GCF_002209165.2',\n",
       " 'GCF_024204785.1',\n",
       " 'GCF_030220105.1',\n",
       " 'GCF_021513115.1',\n",
       " 'GCF_021249285.1',\n",
       " 'GCF_002633805.1',\n",
       " 'GCF_032077755.1',\n",
       " 'GCF_022405475.1',\n",
       " 'GCF_003856455.1',\n",
       " 'GCF_030534115.1',\n",
       " 'GCF_025559465.1',\n",
       " 'GCF_024741415.1',\n",
       " 'GCF_003966585.1',\n",
       " 'GCF_020886635.1',\n",
       " 'GCF_014490515.1',\n",
       " 'GCF_900635305.1',\n",
       " 'GCF_003073395.1',\n",
       " 'GCF_001432245.1',\n",
       " 'GCF_015219885.1',\n",
       " 'GCF_024741595.1',\n",
       " 'GCF_035907395.1',\n",
       " 'GCF_016904135.1',\n",
       " 'GCF_022405495.1',\n",
       " 'GCF_014962425.1',\n",
       " 'GCF_034032775.1',\n",
       " 'GCF_030717625.1',\n",
       " 'GCF_026625225.1',\n",
       " 'GCF_002442895.1',\n",
       " 'GCF_017161095.1',\n",
       " 'GCF_009912015.1',\n",
       " 'GCF_009728775.1',\n",
       " 'GCF_024734775.1',\n",
       " 'GCF_024363445.1',\n",
       " 'GCF_002000725.1',\n",
       " 'GCF_000237125.2',\n",
       " 'GCF_029204095.1',\n",
       " 'GCF_031190615.1',\n",
       " 'GCF_002097595.2',\n",
       " 'GCF_018336515.1',\n",
       " 'GCF_009739865.1',\n",
       " 'GCF_000756205.1',\n",
       " 'GCF_013467405.1',\n",
       " 'GCF_034042655.1',\n",
       " 'GCF_022405935.1',\n",
       " 'GCF_008728795.1',\n",
       " 'GCF_025560945.1',\n",
       " 'GCF_900607285.1',\n",
       " 'GCF_030435955.1',\n",
       " 'GCF_030520085.1',\n",
       " 'GCF_030533955.1',\n",
       " 'GCF_019551355.1',\n",
       " 'GCF_034022875.1',\n",
       " 'GCF_016026635.1',\n",
       " 'GCF_900324245.1',\n",
       " 'GCF_021398305.1',\n",
       " 'GCF_023612155.1',\n",
       " 'GCF_901542335.1',\n",
       " 'GCF_024762175.1',\n",
       " 'GCF_024762195.1',\n",
       " 'GCF_020827015.1',\n",
       " 'GCF_033037535.1',\n",
       " 'GCF_024206615.1',\n",
       " 'GCF_033036395.1',\n",
       " 'GCF_022936265.1',\n",
       " 'GCF_000250905.1',\n",
       " 'GCF_900475805.1',\n",
       " 'GCF_016549375.1',\n",
       " 'GCF_900619585.1',\n",
       " 'GCF_028471605.1',\n",
       " 'GCF_903890635.1',\n",
       " 'GCF_016027715.1',\n",
       " 'GCF_900474455.1',\n",
       " 'GCF_001896085.1',\n",
       " 'GCF_000494895.1',\n",
       " 'GCF_015689355.1',\n",
       " 'GCF_034422215.1',\n",
       " 'GCF_900478185.1',\n",
       " 'GCF_900474815.1',\n",
       " 'GCF_033096385.1',\n",
       " 'GCF_019774635.1',\n",
       " 'GCF_014334715.1',\n",
       " 'GCF_031767395.1',\n",
       " 'GCF_900474825.1',\n",
       " 'GCF_003994915.1',\n",
       " 'GCF_000210955.1',\n",
       " 'GCF_023499275.1',\n",
       " 'GCF_012220405.1',\n",
       " 'GCF_002238315.1',\n",
       " 'GCF_024206755.1',\n",
       " 'GCF_014041875.1',\n",
       " 'GCF_001552035.1',\n",
       " 'GCF_000344765.1',\n",
       " 'GCF_002804185.1',\n",
       " 'GCF_000756485.1',\n",
       " 'GCF_019856275.1',\n",
       " 'GCF_019856395.1',\n",
       " 'GCF_024206635.1',\n",
       " 'GCF_900475565.1',\n",
       " 'GCF_005164885.1',\n",
       " 'GCF_003939065.1',\n",
       " 'GCF_011290565.1',\n",
       " 'GCF_002947575.1',\n",
       " 'GCF_013488145.1',\n",
       " 'GCF_011387195.1',\n",
       " 'GCF_015291825.1',\n",
       " 'GCF_022670815.1',\n",
       " 'GCF_000785515.1',\n",
       " 'GCF_022069505.1',\n",
       " 'GCF_900693055.1',\n",
       " 'GCF_005160785.1',\n",
       " 'GCF_013307185.1',\n",
       " 'GCF_900475085.1',\n",
       " 'GCF_002197265.1',\n",
       " 'GCF_017359885.1',\n",
       " 'GCF_033167795.1',\n",
       " 'GCF_036347595.1',\n",
       " 'GCF_023704295.1',\n",
       " 'GCF_024334065.1',\n",
       " 'GCF_000236475.1',\n",
       " 'GCF_016888845.1',\n",
       " 'GCF_004114735.1',\n",
       " 'GCF_900636445.1',\n",
       " 'GCF_025562855.1',\n",
       " 'GCF_010993845.2',\n",
       " 'GCF_021496885.1',\n",
       " 'GCF_000211075.1',\n",
       " 'GCF_949788935.1',\n",
       " 'GCF_001729925.2',\n",
       " 'GCF_000306805.1',\n",
       " 'GCF_018219175.1',\n",
       " 'GCF_900475045.1',\n",
       " 'GCF_035918475.1',\n",
       " 'GCF_004010855.1',\n",
       " 'GCF_000817065.1',\n",
       " 'GCF_002078855.1',\n",
       " 'GCF_012220065.1',\n",
       " 'GCF_000253315.1',\n",
       " 'GCF_033145555.1',\n",
       " 'GCF_001623565.1',\n",
       " 'GCF_033096305.1',\n",
       " 'GCF_014191835.1',\n",
       " 'GCF_946902905.1',\n",
       " 'GCF_025739385.1',\n",
       " 'GCF_020844005.1',\n",
       " 'GCF_020542025.1',\n",
       " 'GCF_029278525.1',\n",
       " 'GCF_009866825.1',\n",
       " 'GCF_009738105.1',\n",
       " 'GCF_022318505.1',\n",
       " 'GCF_000231885.1',\n",
       " 'GCF_029691405.1',\n",
       " 'GCF_015689415.1',\n",
       " 'GCF_015839415.1',\n",
       " 'GCF_008728875.1',\n",
       " 'GCF_016724885.1',\n",
       " 'GCF_002736265.1']"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(selected_genomes[[\"ftp_path\", \"#assembly_accession\"]]['#assembly_accession'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "id": "00816714-c7c2-4d24-8776-16bf4e78a472",
   "metadata": {},
   "outputs": [],
   "source": [
    "to_download = []\n",
    "\n",
    "for species in list(selected_genomes['ftp_path']): \n",
    "    to_download.append(f'{species}/{species.split(\"/\")[-1]}_genomic.fna.gz')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "id": "1c3061f4-c7c8-4084-ae1c-bd2515c4af32",
   "metadata": {},
   "outputs": [],
   "source": [
    "def save_list_to_file(lst, filename):\n",
    "    with open(filename, 'w') as f:\n",
    "        for item in lst:\n",
    "            f.write(str(item) + '\\n')\n",
    "    print(f\"List has been saved to {filename}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "id": "85344998-7300-44f2-83ad-4be5a6bed62a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "List has been saved to to_download.txt\n"
     ]
    }
   ],
   "source": [
    "save_list_to_file(to_download, 'to_download.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "id": "500f0a0c-de22-4ca3-ab0c-ee75bbd21861",
   "metadata": {},
   "outputs": [
    {
     "ename": "AttributeError",
     "evalue": "'list' object has no attribute 'to_csv'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mAttributeError\u001b[0m                            Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[73], line 1\u001b[0m\n\u001b[0;32m----> 1\u001b[0m \u001b[43mto_download\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mto_csv\u001b[49m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mdownload_genomes.txt\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[0;31mAttributeError\u001b[0m: 'list' object has no attribute 'to_csv'"
     ]
    }
   ],
   "source": [
    "to_download.to_csv('download_genomes.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4e5c2e90-f75c-4c6f-a5ee-2e77f437f600",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "e49c0259-fdbd-49cf-b89b-4177792e82e0",
   "metadata": {},
   "source": [
    "### Generate fastq files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7889681d-5c13-469a-9a00-ebaf28bf1640",
   "metadata": {},
   "outputs": [],
   "source": [
    "import random\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "\n",
    "def generate_microbiome_sample(genome_files, composition, read_length, num_reads):\n",
    "    microbiome_reads = []\n",
    "\n",
    "    for genome_file, abundance in composition.items():\n",
    "        # Load the genome\n",
    "        sequences = SeqIO.parse(genome_file, \"fasta\")\n",
    "        genome_reads = []\n",
    "        \n",
    "        # Generate reads from the genome\n",
    "        for sequence in sequences:\n",
    "            seq_length = len(sequence.seq)\n",
    "            if seq_length >= read_length:\n",
    "                for _ in range(int(abundance * num_reads)):\n",
    "                    start_position = random.randint(0, seq_length - read_length)\n",
    "                    read_seq = sequence.seq[start_position:start_position+read_length]\n",
    "                    read_record = SeqRecord(read_seq, id=f\"{sequence.id}_read\", description=f\"Generated read from {sequence.id}\")\n",
    "                    genome_reads.append(read_record)\n",
    "\n",
    "        # Randomly shuffle reads to avoid sequence bias\n",
    "        random.shuffle(genome_reads)\n",
    "\n",
    "        # Add reads to microbiome sample\n",
    "        microbiome_reads.extend(genome_reads[:int(abundance * num_reads)])\n",
    "\n",
    "    # Randomly shuffle all the selected reads\n",
    "    random.shuffle(microbiome_reads)\n",
    "\n",
    "    return microbiome_reads\n",
    "\n",
    "# Example usage:\n",
    "genome_files = {\n",
    "    'genome1.fasta': 0.2,  # Path to the genome file and its abundance in the composition\n",
    "    'genome2.fasta': 0.3,\n",
    "    # Add more genome files and their abundances as needed\n",
    "}\n",
    "\n",
    "composition = {\n",
    "    'genome1.fasta': 0.5,  # Abundance of each genome in the sample\n",
    "    'genome2.fasta': 0.5,\n",
    "    # Add more genomes and their abundances as needed\n",
    "}\n",
    "\n",
    "read_length = 300  # Length of each read\n",
    "num_reads = 1000  # Total number of reads in the synthetic microbiome sample\n",
    "\n",
    "microbiome_reads = generate_microbiome_sample(genome_files, composition, read_length, num_reads)\n",
    "\n",
    "# Now you have a list of SeqRecord objects containing the synthetic microbiome reads\n",
    "# You can write them to a new fasta file using SeqIO.write\n",
    "SeqIO.write(microbiome_reads, \"synthetic_microbiome.fasta\", \"fasta\")\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
